### ISSUE TITLE
[BUG] cifs share as source fatals because of weird unreadable link as an early-seen file

### ISSUE THREAD

**Issue by tecordes** [rdiff-backup/rdiff-backup#817]
## Bug summary

I have a bug I've traced and solved, but think it should be fixed in the official sources.

Backing up a cifs-share mounted onto a linux box ("B").  The cifs share is the C:\Users dir on a Windows 7 system ("A").  That dir contains a fubar'd entry for "All Users" that looks like this on the linux box:
l---------  1 netbak root    0 Jul 14  2009 All Users -> /??/C:/ProgramData

That file is unreadable, unopenable, un-everythingable.  I just want it skipped.  In fact, I am purposely excluding it with --exclude options, and it's not even in the --include path.

Running rdiff-backup from another linux box ("C") (as the destination) via ssh gives me:

WARNING: ListError: '(b'All Users',)' [Errno 2] No such file or directory: b'/mnt/netbak/All Users'
1.38KiB [1.42KiB/s]
Traceback (most recent call last):
  File "/bin/rdiff-backup", line 33, in <module>
    sys.exit(load_entry_point('rdiff-backup==2.2.0', 'console_scripts', 'rdiff-backup')())
  File "/usr/lib64/python3.10/site-packages/rdiffbackup/run.py", line 37, in main
    sys.exit(main_run(sys.argv[1:]))
  File "/usr/lib64/python3.10/site-packages/rdiffbackup/run.py", line 99, in main_run
    ret_val |= conn_act.setup()
  File "/usr/lib64/python3.10/site-packages/rdiffbackup/actions/backup.py", line 112, in setup
    self.repo.base_dir.conn.fs_abilities.backup_set_globals(
  File "/usr/lib64/python3.10/site-packages/rdiff_backup/fs_abilities.py", line 1076, in backup_set_globals
    src_fsa = rpin.conn.fs_abilities.get_readonly_fsa('source', rpin)
  File "/usr/lib64/python3.10/site-packages/rdiff_backup/connection.py", line 553, in __call__
    return self.connection.reval(*(self.call_name, ) + args)
  File "/usr/lib64/python3.10/site-packages/rdiff_backup/connection.py", line 432, in reval
    raise result
FileNotFoundError: [Errno 2] No such file or directory: b'/mnt/netbak/All Users'

Running a local test instance on box B with the actual mount on it gives better detail:

WARNING: ListError: '(b'All Users',)' [Errno 2] No such file or directory: b'/mnt/netbak/All Users'
Traceback (most recent call last):
  File "/bin/rdiff-backup", line 33, in <module>
    sys.exit(load_entry_point('rdiff-backup==2.2.0', 'console_scripts', 'rdiff-backup')())
  File "/usr/lib64/python3.10/site-packages/rdiffbackup/run.py", line 37, in main
    sys.exit(main_run(sys.argv[1:]))
  File "/usr/lib64/python3.10/site-packages/rdiffbackup/run.py", line 99, in main_run
    ret_val |= conn_act.setup()
  File "/usr/lib64/python3.10/site-packages/rdiffbackup/actions/backup.py", line 112, in setup
    self.repo.base_dir.conn.fs_abilities.backup_set_globals(
  File "/usr/lib64/python3.10/site-packages/rdiff_backup/fs_abilities.py", line 1078, in backup_set_globals
    src_fsa = rpin.conn.fs_abilities.get_readonly_fsa('source', rpin)
  File "/usr/lib64/python3.10/site-packages/rdiff_backup/fs_abilities.py", line 1064, in get_readonly_fsa
    return FSAbilities(desc_string, rp, read_only=True)
  File "/usr/lib64/python3.10/site-packages/rdiff_backup/fs_abilities.py", line 70, in __init__
    self._init_readonly()
  File "/usr/lib64/python3.10/site-packages/rdiff_backup/fs_abilities.py", line 142, in _init_readonly
    self._detect_case_sensitive_readonly(self.root_rp)
  File "/usr/lib64/python3.10/site-packages/rdiff_backup/fs_abilities.py", line 377, in _detect_case_sensitive_readonly
    triple = find_letter(rp)
  File "/usr/lib64/python3.10/site-packages/rdiff_backup/fs_abilities.py", line 352, in find_letter
    file_rp = subdir.append(filename)
  File "/usr/lib64/python3.10/site-packages/rdiff_backup/rpath.py", line 1073, in append
    return self.__class__(self.conn, self.base, self.index + ext, None)
  File "/usr/lib64/python3.10/site-packages/rdiff_backup/rpath.py", line 662, in __init__
    self.setdata()
  File "/usr/lib64/python3.10/site-packages/rdiff_backup/rpath.py", line 716, in setdata
    self.conn.rpath.setdata_local(self)
  File "/usr/lib64/python3.10/site-packages/rdiff_backup/rpath.py", line 1942, in setdata_local
    rpath.chmod(0o400 | rpath.getperms())
  File "/usr/lib64/python3.10/site-packages/rdiff_backup/rpath.py", line 721, in chmod
    self.conn.os.chmod(self.path,
FileNotFoundError: [Errno 2] No such file or directory: b'/mnt/netbak/All Users'

There are a couple of problems.

1. get_readonly_fsa eventually calls rpath.setdata which actually does "write" stuff on this source(!) filesystem, hence the chmod which fails, because it points to a bogus file and the chmod is being done on the referent, not the symlink inode.  In any event, no write should be done in the fs_abilities read-only checks.  The cifs mount in my case is mounted readonly and writes will always fail.

2. find_letter is looking for files in the base path and not honoring --include and --excludes, or even special-file options.  I do the method of setting base to /mnt/netbak, then doing --exclude /mnt/netbak, and then --include for the sub path I actually want (say "/mnt/netbak/Trevor"), which doesn't even include "All Users".  I think the source fs_abilities should limit itself to checking dirs/files that pass the include/exclude rules.  Even if they obeyed the --exclude-special-files it would not have this bug.

3. The code path in the trace does do lstat checks earlier, yet farther down the trace it tries to do things on the dereferenced symlink: the whole path should operate on the symlink itself, or symlinks should simply be excluded for these types of fs_ability checks.

4. It would be nice if there was an option to exclude completely fubar files that cifs mounts of Windows can produce, like this "All Users".  I can't seem to find any mount.cifs options that would fix this, so it has to be done in the software.  The "link" seems fine on the Windows box itself.

My fix that is working for me is at bottom.

## Version, Python, Operating System

---
exec:
  api_version:
    actual: 0
    default: 200
    max: 201
    min: 200
  argv:
  - /bin/rdiff-backup
  - info
  parsed: null
  version: 2.2.0
python:
  executable: /usr/bin/python3
  name: cpython
  version: 3.10.9
system:
  fs_encoding: utf-8
  platform: Linux-6.0.15-200.fc36.x86_64-x86_64-with-glibc2.35
...

## rdiff-backup call

normal case run on box C (don't bother trying to figure it out):

/bin/nice -n19 /usr/bin/ionice -c2 -n7 rdiff-backup  --force --tempdir /data/Tmp --remote-schema 'ssh -x -C -q -o StrictHostKeyChecking=no -o CheckHostIP=no -c aes256-gcm@openssh.com {h} '\''(/usr/local/sbin/netbak-smbmount '\''//192.168.100.2/Users'\'' '\''/mnt/netbak'\'' '\''netbak'\'' '\''0'\'' && ( {{ ok=1;  (( `find /mnt/netbak/Joe/Favorites | head -2 | wc -l` == 2 )) && ( umask 077; chmod --quiet 600 /tmp/rdb.log; rnice=/usr/bin/nice; [ -e /bin/nice ] && rnice=/bin/nice; $rnice -n19 /usr/bin/ionice -c2 -n7 rdiff-backup  server 2>>/tmp/rdb.log ) && ok=0; }} ; /usr/local/sbin/netbak-smbumount /mnt/netbak; rm -f /home/netbak/.netbak-smbmount.cred; exit $ok )) | /usr/bin/pv -L 200000 -rbWf -i2592000'\' backup  --include '/mnt/netbak/Joe/Favorites' --exclude '/mnt/netbak' --exclude '/mnt/netbak/All Users' --exclude-special-files 'netbak@mydomain.com::/mnt/netbak' /data/Netbak/mydomain.com/joes/Users/JoeFavorites 2>&1

the test case on box B that better illustrates the bug:

rdiff-backup backup --include '/mnt/netbak/Joe/Favorites' --exclude '/mnt/netbak' --exclude-regexp 'All' --exclude '/mnt/netbak/All Users' --exclude-special-files /mnt/netbak /tmp/R

The source dir (mounted from box A) looks like (on box B):
4:14am ~#ll /mnt/netbak
total 13
drwxr-xr-x  2 netbak root 4096 Apr 13  2018 ./
drwxr-xr-x. 4 root   root 4096 Aug  9 08:57 ../
l---------  1 netbak root    0 Jul 14  2009 All Users -> /??/C:/ProgramData
drwxr-xr-x  2 netbak root    0 Jul  5  2022 Joe/
...

## What happened and what did you expect?

Fatal error:
FileNotFoundError: [Errno 2] No such file or directory: b'/mnt/netbak/All Users'

## More information

My ugly fix was to duplicate in rpath.py def append into, with minor edits:
```python
    def appendnodata(self, *ext):
        return self.__class__(self.conn, self.base, self.index + ext, None, True)
```
and adjust class RPath __init__ to:
```python
    def __init__(self, connection, base, index=(), data=None, ignoredata=None):
```
and edit the if data is None line to:
```python
            if data is None and ignoredata is None:
```
Then in fs_abilities.py edit find_letter and test_triple to call appenddata instead of append.

Now it works great.  I still get a:
WARNING: ListError: '(b'All Users',)' [Errno 2] No such file or directory: b'/mnt/netbak/All Users'
but I can filter that out.  It would be nice if my ideas #2 or #4 above were done then this error wouldn't occur at all (as it's not something that gets backed up anyhow).

Thank you very much!

**Comment by ericzolf** [comment]
Option 2 would require a complete restructuration of the code, how it handles exclude/include, it probably ain't going to happen any time soon. What I think could be done, is to add an option `--ignore` instead of exclude, which would basically cover your points 2 and 4.

Regarding your patch, again, could you provide it in a more readable format?

I added for you the necessary code, as in:

```
```python
def __init__()
  pass
```_
```

**Comment by ericzolf** [comment]
I acknowledge the bug but I don't think I like the fix, too specific, it should be more generic (but would introduce greater changes). Need to think a bit more about it.

**Comment by tecordes** [comment]
I completely agree.  My fix is an ugly hack as they don't address the real problem(s).  I found the code path too hard to follow without better knowledge of the paradigm behind all those sub-calls and their use elsewhere throughout the code.  I can see the idea of code re-use here, and didn't want to muck anything up for the other callers.

I think my (1) point might be good enough to be its own bug, and is not required to fix the problem.  Clearly a call path with "read only" in the function names should not be writing stuff deep down by accident.

I hoped (2) would be easy for you if the code that handles those things could be reused for this.  But I completely understand that if it's not structured that way then making it obey inc/exc would be arduous.  In theory, I suppose it does make sense that a fs_ability check should be able to just run/apply to the fs as a whole, regardless of whether it's inside or outside the inc/exc parameters.  But in practice...?

So maybe a mix of (1) and (3) is the right approach.  (1) make sure the read-only's are never trying to write; and (3) fs_abilities should never run its (non-symlink) checks on symlinks?

Or, as a novel idea, and maybe easiest of all, there could be a new option (just for people experiencing problems to enable) like --abilities-path where the user could hardcode the path that all fs_abilities checks are performed upon.  Then the abilities stuff limits itself to operating there.  Doesn't really fix the bug for the unaware, though.

**Comment by ericzolf** [comment]
Two things:

1. can you tell me how you export resp. import your CIFS share? Perhaps I can reproduce the setup, that would make tests easier.
2. can you please check if `os.lchmod(b'All Users', 0o400)` works? (notice the l for link)

**Comment by ericzolf** [comment]
@tecordes reiterating my questions...

**Comment by ericzolf** [comment]
OK, I have my reproducer:

```
>net share users=C:\Users
$ sudo mount -t cifs -o username=vagrant,password=xxx,uid=1000,gid=1000 //192.168.121.123/users /mnt
$ mount
[...]
//192.168.121.148/users on /mnt type cifs (rw,relatime,vers=3.1.1,cache=strict,username=vagrant,uid=1000,noforceuid,gid=1000,noforcegid,addr=192.168.121.123,file_mode=0755,dir_mode=0755,soft,nounix,serverino,mapposix,rsize=4194304,wsize=4194304,bsize=1048576,echo_interval=60,actimeo=1,closetimeo=5)
(rdb) [ericl@t460s rdiff-backup]$ ll /mnt
total 1
l---------. 1 ericl ericl   0 Sep 15  2018 'All Users' -> '/??/C:/ProgramData'
dr-xr-xr-x. 2 ericl ericl   0 May 17  2022  Default
drwxr-xr-x. 2 ericl ericl   0 Sep 15  2018 'Default User'
-rwxr-xr-x. 1 ericl ericl 174 Sep 15  2018  desktop.ini
dr-xr-xr-x. 2 ericl ericl   0 May 17  2022  Public
drwxr-xr-x. 2 ericl ericl   0 Jan 26 06:31  vagrant
$ rdiff-backup backup --include '/mnt/vagrant/Favorites' --exclude '/mnt' --exclude-regexp '.*' --exclude '/mnt/All Users' --exclude-special-files /mnt /tmp/bak
$ rdiff-backup backup --exclude-special-files --include '/mnt/vagrant/Favorites' --exclude '/mnt' /mnt /tmp/bak
```

Note that the excludes after the `/mnt` one are redundant, independently from the current issue, and the 2nd rdiff-backup command would be more correct.

**Comment by ericzolf** [comment]
I didn't realize that lchown exists only under BSD & Co (at least, neither Linux nor Windows), so not an option either.

But this is probably the main reason for this bug: normally symlinks are fully readable so this part of the code never kicks in, as it does only when the file's attributes can't be read. OK, I'll work on a fix now.
